File: CODE/BilexiconUtil.m
classdef BilexiconUtil
   
    methods(Static)
        %% FUNCTIONS TO LOAD THE BILEXICON
        function lex = new()
            lex.s2t = java.util.Hashtable; % source to target
            lex.t2s = java.util.Hashtable; % target to source
        end
        
        function lex = load(filename)
            load(filename); % loading B;
            lex = BilexiconUtil.parseLexicon(B);
        end
        
        function lex = parseLexicon(B)
            N = size(B,1);
            lex = BilexiconUtil.new();
            for n=1:N, % go over words.
                W1 = B{n,1};
                W2 = B{n,2};
                
                if length(W1)==0 || length(W2)==0 % skip empty values.
                    continue
                end
                % map source words to a list of target words and vice
                % versa.
    
                lex.s2t = BilexiconUtil.listadd(lex.s2t, W1, {W2});
                lex.t2s = BilexiconUtil.listadd(lex.t2s, W2, {W1});
            end
        end
        
        %% GENERAL HASH OF LIST FUNCTIONS
        
        function h = listadd(h, key, values)
            key = BilexiconUtil.fixWord(key);
            list = h.get(key);
            if isempty(list)
                h.put(key, java.util.LinkedList);
            end
            for i = 1:length(values),
                word = values{i};
                word = BilexiconUtil.fixWord(word);
                h.get(key).add(word);
            end
        end
        
        function word = fixWord(word)
            if ~strcmp(word(1), '_') % words are padded with starting and trailing '_'.
                word = ['_',word,'_'];
            end
        end

        function hash = new_addAll(list)
            % yes, I know there's a function addAll
            hash = java.util.HashSet;
            N = size(list,1);
            for n=1:N,
                hash.add(list{n});
            end
        end
        
        %% SET UP THE BILEXICON
        
        %function gtlex = ground_truth(lex, source_words, target_words, seed_source_words, seed_target_words)
        function gtlex = ground_truth(lex, source_words, target_words)
            % This function prepares the ground truth bilexicon - gtlex
            % scoring will be done against gtlex.
            % gtlex is a sub-set of lex, restricted to the given
            % source and target words.
            
            % setup a hash for quick lookup
            HS.source = BilexiconUtil.new_addAll(source_words);
            HS.target = BilexiconUtil.new_addAll(target_words);
    
            % remove words from the bilexicon lex.s2t/t2s if they don't
            % appear in the source/target.
            gtlex.s2t = BilexiconUtil.remove_missing_words(lex.s2t, HS.source, HS.target);
            gtlex.t2s = BilexiconUtil.remove_missing_words(lex.t2s, HS.target, HS.source);
        end

        function bilex = remove_existing_words(bilex, keys_to_remove)
            % removes a set of words from the hash
            keys_to_remove = keys_to_remove.toArray();
            for i=1:keys_to_remove.size(),
                key = keys_to_remove(i);
                if bilex.containsKey(key)
                    bilex.remove(key);
                end
            end
        end
        
        function new_bilex = remove_missing_words(bilex, sourceHS, targetHS)
            % In a clean hash, put only keys that are present in sourceHS
            % and, remove target words not in targetHS.
            % so that only possible edges can be counted
            source_words = sourceHS.toArray();
            new_bilex    = java.util.Hashtable;
            for i=1:source_words.size(),
                source_word = source_words(i);
                if bilex.containsKey(source_word)
                    target_list = bilex.get(source_word);
                    % remove words not in target
                    target_itr = target_list.listIterator();
                    while target_itr.hasNext()
                        target_word = target_itr.next();
                        if ~targetHS.contains(target_word)
                            target_itr.remove();
                        end
                    end
                    if target_list.size() > 0
                        new_bilex.put(source_word, target_list);
                    end
                end
            end
        end
        
        %% scoring
        
        function score = F1(precision, recall, beta)
            if nargin < 3
                beta = 1;
            end
            denom = beta^2*precision + recall;
            score = (1+beta^2).*(precision.*recall)./denom;
        end
        
        function scores = getF1scores(gtlex, matching, weights, options)
            % returns precision, recall and F1 scores.
            minWeight = min(weights);
            maxWeights = max(weights);
            %[matching, mat2cell(weights', ones(length(weights),1))]
            
            M = gtlex.s2t.size(); % we use the s2t mapping as the reference (practically s should be English, for which words are easy to obtain).
            % sort matching in ascending order.
            N = size(matching,1);
            assert(length(weights) == N);
            [weights,pi] = sort(weights);
            matching = matching(pi, :);

            C = zeros(N,3); 
            for n=1:N,
                source_word = matching{n,1};
                target_word = matching{n,2};
                C(n,1) = 1; % count size of matching being considered
                if gtlex.s2t.containsKey(source_word) % some match exists for this source word
                    C(n,2) = 1; % counts number of existing word in source and lexicon
                    C(n,3) = BilexiconUtil.is_valid_s2t_match(gtlex, source_word, target_word); % counts valid matches
                end
            end
            C = cumsum(C);
            
            % C1 - always 1, i.e. cumsum counts the number of words.
            % C2 - 1 of source
            % C1 - counts the size of the matching being considered
            
            scores.M = M;
            scores.precision = C(:,3) ./ C(:,2);
            scores.recall = C(:,3) ./ M;
            scores.F1 = BilexiconUtil.F1(scores.precision,scores.recall);
            scores.N = C(:,2);
            scores.R = C(:,1);
        end
        
        function b = is_valid_s2t_match(gtlex, source_word, target_word)
            b = 0;
            target_list = gtlex.s2t.get(source_word);
            if ~isempty(target_list)
                target_words = target_list.toArray();
                for i=1:target_list.size()
                    word = target_words(i);
                    if strcmpi(word, target_word)
                        b = 1;
                        break;
                    end   
                end
            end
        end
        
        %% OUTPUT
        
        function outputScores(scores, options, title)
            fprintf('Results for %s:\n', title);
            fprintf('===============\n');
            %options
            % find precisions based on recall levels
            i05 = find(scores.recall>=0.05,1);
            i10 = find(scores.recall>=0.1,1);
            i20 = find(scores.recall>=0.20,1);
            i25 = find(scores.recall>=0.25,1);
            i33 = find(scores.recall>=0.33,1);
            i40 = find(scores.recall>=0.40,1);
            i50 = find(scores.recall>=0.50,1);
            i60 = find(scores.recall>=0.60,1);
            
            I = unique([i05, i10, i20, i25, i33, i40, i50, i60]);
            
            N  = scores.R(I);
            P  = scores.precision(I);
            R  = scores.recall(I);
            F1 = scores.F1(I);
            fprintf('Ra'); fprintf('%8.2f',100*N/max(scores.R)); fprintf('\n');
            fprintf('F1'); fprintf('%8.2f', 100*F1); fprintf('\n');
            fprintf('Pr'); fprintf('%8.2f', 100*P);  fprintf('\n');
            fprintf('Re'); fprintf('%8.2f', 100*R);  fprintf('\n');
            fprintf('=======\n');
            if ~strcmpi(title, 'Edit Distance')
                fprintf('Params: [wtype="%s", lambda=%2.2f, M=%d, K=%2.2f]\n', options.weight_type, options.lambda, options.M, options.K);
                fprintf('=======\n');
            end
        end
        
        %%%%%%%%%%%%%%%%% TEST FUNCTIONS
           
        
        function gtlex = test()
            % test function for this class.
            B = {
                '_big_', '_grande_';
                '_the_', '_el_';
                '_the_', '_la_';
                '_the_', '_un_';
                '_thin_', '_delgado0_';
                '_thin_', '_delgado1_';
                '_city_', '_ciudad_';
                '_country_', '_pais_';
                '_beautiful_', '_hermosa_';
                };
            
            % build the lexicon from B.
            bilex = BilexiconUtil.parseLexicon(B);
            
            matching = {
                '_big_', '_delgado1_';
                '_city_', '_el_';
                '_thin_', '_delgado0_';
                '_country_', '_hermosa_';
                '_beautiful_', '_pais_';
                '_unmatched_', '_unmatched_'
                };
            
            weights = 1:size(matching,1); % just some simple weights
            
            % filter with ground truth.
            gtlex  = BilexiconUtil.ground_truth(bilex, matching(:,1), matching(:,2));
            scores = BilexiconUtil.getF1scores(gtlex, matching, weights);
            plot(scores.F1);
            
            BilexiconUtil.outputScores(scores, [], 'test run');
        end
    end
    
end

File: newcode/BilexiconUtil.py
# computes precision and recall scores
from collections import defaultdict
import math
import common
import sys
import csv
import perm
import numpy as np


def readLexicon(filename, delimiter='\t'):
    print >> sys.stderr, 'reading Bilexicon:', filename
    dict = defaultdict(set)
    j = 0
    with open(filename, 'rb') as csvfile:
        reader = csv.reader(csvfile, delimiter=delimiter, quotechar='|')
        for row in reader:
            source_word = row[0]  # .lower()
            for target_word in row[1:]:
                #dict[source_word].add(target_word.lower())
                dict[source_word].add(target_word)
            j += 1
    print >> sys.stderr, 'Done reading Bilexicon'
    return dict


# dictionary contains the list of all matching obtained from the gold lexicon.
# dictionary has format dict of key, to dict of values
# source words is a dictionary of source words and target words is a dictionary of target words
def filterLexicon(lex, source_words, target_words):
    F = defaultdict(set)
    total_items = 0
    # keep only pairs (u,w) from dictionary that are both in the source and target lists.
    for u in lex:
        if u not in source_words:
            continue  # u is not among the source words
        for w in lex[u]:
            if w in target_words:
                F[u].add(w)
                total_items += 1
    return F, total_items


# filtered dictionary is a dictionary with the key as a source word and the value
# is another dictionary with the key as the target words
# matches is a list of lists. Each row in the list has the format
# [source_word,matched_target_word,score]. The items are sorted according to score
def getScores(lex, source_words, target_words, weights):
    M = len(lex)
    # sort according to weights (increasing)
    (_, pi) = perm.sort(weights, reverse=False)
    source_words = source_words[pi]
    target_words = target_words[pi]
    N = len(source_words)
    assert N == len(target_words)
    C = np.zeros((N, 3))  # [1, exists in source, target matches]
    dict_keys = lex.keys()
    for i, (source_word) in enumerate(source_words):
        target_word = target_words[i]
        C[i, 0] = 1  # always 1
        if source_word in dict_keys:
            C[i, 1] = 1  # word exists as a source word
            #if target_word in lex[source_word]:
            if is_valid_match(lex, source_word, target_word):
                C[i, 2] = 1  # (source, target) words are correctly matched according to dict

    C = np.cumsum(C, 0)  # cumulative sum per column
    scores = common.Struct()
    scores.M = M
    scores.precision = C[:, 2] / C[:, 1]
    scores.recall = C[:, 2] / M
    scores.F1 = F1(scores.precision, scores.recall)

    return scores  # C should allow computing precision/recall/F1 for any cutoff value.


def is_valid_match(lex, source_word, target_word):
    return target_word in lex[source_word]


def format(x, format='.3f'):
    if np.isscalar(x):
        return ('{:'+format+'}').format(x)
    else:
        return np.array([('{:'+format+'}').format(i) for i in x])


def outputScores(scores, title):
    cutoff = [0.1, 0.25, 1.0/3, 0.5, 2.0/3, 0.75, 0.77, 0.8]
    p = []
    r = []
    f = []

    d = 3
    cutoff = np.around(cutoff, decimals=d)
    scores.precision = np.around(scores.precision, decimals=d)
    scores.recall = np.around(scores.recall, decimals=d)
    scores.F1 = np.around(scores.F1, decimals=d)

    for i, c in enumerate(cutoff):
        J = np.argwhere(scores.recall >= c)

        if len(J) > 0:
            j = J[0]
            p.append(scores.precision[j][0])
            r.append(scores.recall[j][0])
            f.append(scores.F1[j][0])
        else:
            p.append(0)
            r.append(0)
            f.append(0)

    print 'Final Scores', title, ':'
    print '============='
    print 'Prec''  :', np.array(p)
    print 'Recall:', np.array(cutoff)
    #print 'Recall:', np.array(r)
    print 'F1    :', np.array(f)

    (maxF1, i) = common.max(scores.F1)
    precision_i = scores.precision[i]
    recall_i = scores.recall[i]
    print 'max F1:', (maxF1), 'precision: ', (precision_i), 'recall:', (recall_i)


# computes the F1 score
def F1(p, r, beta=1):
    sqr_beta = beta**2
    D = sqr_beta*p + r
    score = (1 + sqr_beta) * (p*r) / D
    score = [0 if math.isnan(s) else s for s in score]  # fix nans
    return score


if __name__ == '__main__':
    B = defaultdict(defaultdict)
    B['_big_']['_grande_'] = 1
    B['_the_']['_el_'] = 1
    B['_the_']['_la_'] = 1
    B['_the_']['_un_'] = 1
    B['_thin_']['_delgado0_'] = 1
    B['_thin_']['_delgado1_'] = 1
    B['_city_']['_ciudad_'] = 1
    B['_country_']['_pais_'] = 1
    B['_beautiful_']['_hermosa_'] = 1


    matching = []
    matching.append(['_big_', '_delgado1_'])
    matching.append(['_city_', '_el_'])
    matching.append(['_thin_', '_delgado0_'])
    matching.append(['_country_', '_hermosa_'])
    matching.append(['_beautiful_', '_pais_'])
    matching.append(['_unmatched_', '_unmatched_'])
    matching = np.array(matching)

    S = matching[:, 0]
    T = matching[:, 1]

    weights = xrange(len(matching)) # just some simple weights
            
    # filter with ground truth.
    gt, total_items = filterLexicon(B, S, T)
    scores = getScores(gt, S, T, weights)

    outputScores(scores,  'test run')
